localstack/services/moto.py

```diff
@@ -65,7 +69,22 @@ def call_moto_with_request(
         region=context.region,
     )
 
-    local_context.request.headers.update(context.request.headers)
+    if override_headers:
+        headers = copy.deepcopy(context.request.headers)
+        # remove the headers that are parameters from the original request
+        header_to_remove = list(
+            header
+            for header in headers.keys()
+            if header.startswith("x-amz") and not header.startswith("x-amz-meta")
+        )
+        for header in header_to_remove:
+            headers.remove(header)
+
+        headers.update(local_context.request.headers)
+
+        local_context.request.headers = headers
+    else:
+        local_context.request.headers.update(context.request.headers)
```

Comment:
I will explain in comments as well. 

What's happening when we use `call_moto_with_request` is that we're re-serializing the parameters of the request.
Amazon uses custom headers as parameters often by them starting by `x-amz-`. 

But when we call moto with the request, it's because we want to override certain parameters for that request.

However, from the order of the method, we first re-serialized the parameters from the action request (the serialized dict), and then we update the headers from the original request, which means you will override your previously set parameter with the one from the original request.
Ex: you have a parameter call `ParameterOne`, located in `x-amz-parameter-one`. Its value is `foo` but you want to pass it to moto with `bar`. You first take the original request dict, set it like `request["ParameterOne"] = "bar"`, serialize it, so you have a proper request with `x-amz-parameter-one=bar` but then you override the headers and end up with `x-amz-parameter-one=foo` again.

This `override_headers` argument allows you to remove those from the original request so that you can properly override everything with the request object you passed.

Reply:
Also, call to `call_moto_with_request` are rather rare in our code base, which is why I took the liberty to introduce this non-breaking change. If I should add a new `call_moto_x`, it's also good for me!

Reply:
Ah, okay, thanks for the explanation! Unfortunately, it's a bit dangerous to expect that all headers have the `x-amz` prefix. The casing is also differs across the different service specs, which could cause problems.
Is there another way to maybe reverse the order, and set the headers first, but then overwrite them with the serialized parameters?

Reply:
Yes, I think that is actually possible. I suppose I got carried away from an initial issue, but by issue `deepcopy`, I could just take the original and use the `local_context` headers to update it then set it to the `local_context`. This should work 👍